File: src/python_ros2_basics/python_ros2_basics/publisher_example.py
#!/usr/bin/env python3

# ros2 python library
import rclpy
# The node class from rclpy
from rclpy.node import Node
# The message that we are using
from std_msgs.msg import String

# Create a class that inherit from ros2 rclpy Node
class TextPublisher(Node):
    def __init__(self):
        # Use the init of the Node class
        super().__init__('text_publisher')
        # Declare a publisher that will publish a String message on the topic
        # example_string_topic, with a queue size of 10
        self.publisher_ = self.create_publisher(String, "example_string_topic", 10)
        
        timer_period = 1.0
        # Create a timer that will call the function self.timer_callback every second
        self.timer = self.create_timer(timer_period, self.timer_callback)
        self.counter = 0
        # logger are a way to print in ROS2 node
        self.get_logger().info('Publisher node has been started.')
    
    # The callback called by the timer
    def timer_callback(self):
        self.counter += 1
        #Create a String message
        msg = String()
        # Set the data field of the string message with our string
        msg.data = f"{self.counter} x " + 'Welcome to Zerobotics Tutorials!'
        # Use the method publish of our the publisher_ to publish message on the topic
        self.publisher_.publish(msg)
        # Print a log message
        self.get_logger().info(f'Published: {msg.data}')

# The main function executed when we are launching the node
def main(args=None):
    # Init rclpy ROS2
    rclpy.init(args=args)
    # Create the node
    node = TextPublisher()
    # Keep the node running
    rclpy.spin(node)
    # Destroy the node on stop
    node.destroy_node()
    # Shutdown the rclpy ROS2
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/cpp_ros2_basics/src/publisher_example.cpp
// Include the necessary C++ headers
#include <memory>
// ROS2 C++ Client library
#include "rclcpp/rclcpp.hpp"
// Include the message type that we will publish
#include "std_msgs/msg/string.hpp"

// Define the TextPublisher class, inheriting from rclcpp::Node
class TextPublisher : public rclcpp::Node {
public:
    TextPublisher() : Node("text_publisher"), counter_(0) {
        // Initialize the publisher, publishing to example_string_topic with a queue size of 10
        publisher_ = this->create_publisher<std_msgs::msg::String>("example_string_topic", 10);

        // Initialize the timer with a 1-second interval
        timer_ = this->create_wall_timer(
            std::chrono::seconds(1),
            std::bind(&TextPublisher::timer_callback, this)
        );

        // Log a message indicating the node has started
        RCLCPP_INFO(this->get_logger(), "Publisher node has been started.");
    }

private:
    // Callback function for the timer
    void timer_callback() {
        counter_++;
        
        // Create a String message
        auto message = std_msgs::msg::String();
        // Fill the field data of our message with a string
        message.data = std::to_string(counter_) + " x Welcome to Zerobotics Tutorials!";

        // Publish the message
        publisher_->publish(message);

        // Log the published message
        RCLCPP_INFO(this->get_logger(), "Published: '%s'", message.data.c_str());
    }

    // Member variables declaration
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
    int counter_;
};

// The main function, equivalent to the Python main
int main(int argc, char * argv[]) {
    // Initialize rclcpp
    rclcpp::init(argc, argv);

    // Create a shared pointer for the TextPublisher node
    auto node = std::make_shared<TextPublisher>();

    // Keep the node spinning
    rclcpp::spin(node);

    // Shutdown and cleanup
    rclcpp::shutdown();
    return 0;
}